## **Project Title: A Visual Exploration of the Relationship Between Trade Tariffs and Germany's Economic Growth from 1988 to 2021**

> Author: Abdullah Akintobi  
> Date: September 19, 2025

### **Problem Statement**

This project conducts an exploratory data analysis of Germany's trade and economic performance from 1988 to 2021 to evaluate the influence of tariff policy changes on key economic indicators.

The central problem is to determine the correlation between Germany's tariff rate fluctuations and its economic growth and trade volumes over the 34-year period.

The analysis will address this problem by answering the following specific research questions:

- **Correlation with Economic Growth:** What is the statistical correlation between changes in Germany's tariff rates (both `AHS Simple Average (%)` and `MFN Simple Average (%)`) and its `Country Growth (%)`?

- **Impact on Trade Volume:** How do changes in tariff policies correspond with fluctuations in Germany's total trade volumes, as measured by `Export (US$ Thousand)` and `Import (US$ Thousand)`?

- **Historical Analysis:** Can distinct periods be identified where a shift in trade policy appears to have a measurable impact on Germany's economic trajectory?

The insights gained will provide empirical evidence of the intricate relationship between a nation's trade policy and economic health, serving as a comprehensive case study.

### **Import Dependencies**

In [13]:
# Import necessary libraries
import numpy as np
import pandas as pd

# Set display options to show all columns
pd.set_option("display.max_columns", None)
pd.set_option("display.width", None)

### **Data Loading and Initial Observations**

In this section, we will:

- Load and preview the raw dataset.

- Select the features to be used for the analysis.

In [14]:
# Load the raw dataset
df_raw = pd.read_csv("../data/raw/34_years_world_export_import_dataset.csv")

# Preview the data
df_raw.head()

Partner Name  Year  Export (US$ Thousand)  Import (US$ Thousand)  \
0        Aruba  1988                3498.10                 328.49   
1  Afghanistan  1988              213030.40               54459.52   
2       Angola  1988              375527.89              370702.76   
3      Anguila  1988                 366.98                   4.00   
4      Albania  1988               30103.56               47709.30   

   Export Product Share (%)  Import Product Share (%)  \
0                     100.0                       100   
1                     100.0                       100   
2                     100.0                       100   
3                     100.0                       100   
4                     100.0                       100   

   Revealed comparative advantage  World Growth (%)  Country Growth (%)  \
0                             NaN               NaN                 NaN   
1                             NaN               NaN                 NaN   
2                             NaN               NaN                 NaN   
3                             NaN               NaN                 NaN   
4                             NaN               NaN                 NaN   

   AHS Simple Average (%)  AHS Weighted Average (%)  AHS Total Tariff Lines  \
0                    2.80                      2.92                   155.0   
1                    0.88                      1.83                   548.0   
2                    2.02                      3.89                   633.0   
3                    3.71                      1.09                    33.0   
4                    1.84                      2.38                   744.0   

   AHS Dutiable Tariff Lines Share (%)  AHS Duty Free Tariff Lines Share (%)  \
0                                18.06                                 60.00   
1                                 8.76                                 82.66   
2                                25.43                                 69.19   
3                                 6.06                                 72.73   
4                                20.83                                 60.48   

   AHS Specific Tariff Lines Share (%)  AHS AVE Tariff Lines Share (%)  \
0                                20.00                            1.94   
1                                 8.03                            0.55   
2                                 5.37                            0.00   
3                                21.21                            0.00   
4                                17.61                            1.08   

   AHS MaxRate (%)  AHS MinRate (%)  AHS SpecificDuty Imports (US$ Thousand)  \
0             50.0              0.0                                  1867.00   
1             35.0              0.0                                 30863.03   
2             40.0              0.0                                723819.51   
3             35.0              0.0                                    60.00   
4             25.0              0.0                                 18806.15   

   AHS Dutiable Imports (US$ Thousand)  AHS Duty Free Imports (US$ Thousand)  \
0                              2346.37                                781.65   
1                             70204.13                              23987.37   
2                            754183.84                             167297.68   
3                                65.00                                518.00   
4                             62294.53                              38901.42   

   MFN Simple Average (%)  MFN Weighted Average (%)  MFN Total Tariff Lines  \
0                   13.59                      8.46                  1152.0   
1                   17.68                     12.43                  4142.0   
2                   12.70                      6.14                  5438.0   
3                   16.63                     14.75                   322.0   
4                   19.20    

In [15]:
# Check the shape of the dataset
df_raw.shape

(8096, 33)

In [16]:
# Get unique values in Partner Name column and sort them alphabetically
unique_partners = sorted(df_raw["Partner Name"].unique())
print(unique_partners)

[' World', 'Afghanistan', 'Albania', 'Algeria', 'American Samoa', 'Andorra', 'Angola', 'Anguila', 'Antarctica', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas, The', 'Bahrain', 'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belgium-Luxembourg', 'Belize', 'Benin', 'Bermuda', 'Bhutan', 'Bolivia', 'Bonaire', 'Bosnia and Herzegovina', 'Botswana', 'Bouvet Island', 'Br. Antr. Terr', 'Brazil', 'British Indian Ocean Ter.', 'British Virgin Islands', 'Brunei', 'Bulgaria', 'Bunkers', 'Burkina Faso', 'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde', 'Cayman Islands', 'Central African Republic', 'Chad', 'Chile', 'China', 'Christmas Island', 'Cocos (Keeling) Islands', 'Colombia', 'Comoros', 'Congo, Dem. Rep.', 'Congo, Rep.', 'Cook Islands', 'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curaçao', 'Cyprus', 'Czech Republic', 'Czechoslovakia', 'Denmark', 'Djibouti', 'Dominica', 'Dominican Republic', 'East Asia & Pacific', 'East Tim

In [17]:
# Print partners that start with 'Ger'
for partner in unique_partners:
    if partner.startswith("Ger"):
        print(partner)

German Democratic Republic
Germany


In [18]:
# Check the years for 'German Democratic Republic'
df_raw[df_raw["Partner Name"] == "German Democratic Republic"]["Year"].unique()

array([1988, 1989, 1990])

**Observations:**

The dataset contains records for both the **[German Democratic Republic](https://en.wikipedia.org/wiki/East_Germany)** (from 1988 to 1990) and **Germany** (from 1988 to 2021). We will combine the records for both countries from 1988 to 1990 into a single unified record for Germany. This is crucial for our analysis because it accurately reflects the political and economic reality of **Germany's reunification in 1990**. Merging these records allows for a **continuous time-series analysis** over the entire 1988-2021 period, preventing a misleading split in the data and providing a more historically representative view of the country's economic trends.

We also observed that the dataset contains 8,096 rows and 33 columns. For the purpose of this project, we will focus on a subset of the data that is essential for our analysis.

Below are the seven key features we will use:

  - **`Partner Name`**: This column will be used to filter the data specifically for 'Germany'.

  - **`Year`**: This time-series variable is essential for plotting data over the 1988-2021 period.

  - **`Country Growth (%)`**: The primary variable for measuring Germany's economic growth.

  - **`AHS Simple Average (%)`**: One of two key tariff variables, representing the "Applied Harmonized System" tariff rate.

  - **`MFN Simple Average (%)`**: The second key tariff variable, representing the "Most Favored Nation" tariff rate.

  - **`Export (US$ Thousand)`**: Used to analyze the volume of trade over time.

  - **`Import (US$ Thousand)`**: Also used to analyze the volume of trade over time.

You can click [here](../data/meta/data_dictionary.csv) to see the full description of each column in the dataset. In the next section we will preprocess the data by perfoming feature selection, removing unwanted records, and merging the record for Germany and German Democratic Republic.

### **Data Preprocessing**

In this section, we will:

- Drop the columns that are not necessary for our analysis.

- Remove the records where Partner Name is not German Democratic Republic or Germany.

- Merge the German Democratic Republic and Germany records as Germany from the years 1988 to 1990.

In [19]:
# Define the columns (features) required for the analysis
FEATURES_TO_KEEP = [
    "Partner Name",
    "Year",
    "Country Growth (%)",
    "AHS Simple Average (%)",
    "MFN Simple Average (%)",
    "Export (US$ Thousand)",
    "Import (US$ Thousand)",
]

# Drop columns that are not necessary for the analysis
df_filtered = df_raw[FEATURES_TO_KEEP]

# Define the countries of interest
countries_of_interest = ["Germany", "German Democratic Republic"]

# Filter the DataFrame to include only records for the countries of interest
df_germany_gdr = df_filtered[
    df_filtered["Partner Name"].isin(countries_of_interest)
].copy()

# Preview the cleaned and filtered DataFrame
df_germany_gdr.head(10)

Partner Name  Year  Country Growth (%)  \
47    German Democratic Republic  1988                 NaN   
48                       Germany  1988                 NaN   
254   German Democratic Republic  1989                 NaN   
255                      Germany  1989                3.78   
464   German Democratic Republic  1990                 NaN   
465                      Germany  1990               12.71   
677                      Germany  1991                6.56   
892                      Germany  1992                2.46   
1124                     Germany  1993               -8.40   
1358                     Germany  1994                5.54   

      AHS Simple Average (%)  MFN Simple Average (%)  Export (US$ Thousand)  \
47                      7.12                   14.39           1.457269e+06   
48                     10.87                   14.06           3.493684e+07   
254                     9.08                   23.22           2.219954e+06   
255                    14.66                   23.62           5.600603e+07   
464                    38.61                   20.13           2.292162e+06   
465                    14.32                   21.52           7.157536e+07   
677                    14.67                   15.62           1.033760e+08   
892                    14.58                   15.84           1.664047e+08   
1124                   12.30                   16.10           1.903617e+08   
1358                   15.87                   18.65           3.157453e+08   

      Import (US$ Thousand)  
47             2.119792e+06  
48             4.149730e+07  
254            2.777570e+06  
255            7.462543e+07  
464            2.934573e+06  
465            9.402324e+07  
677            1.216899e+08  
892            1.904864e+08  
1124           2.096702e+08  
1358           3.487525e+08

In [20]:
# Define the years for merging
YEARS_TO_MERGE = [1988, 1989, 1990]

# Split data: merge years vs keep others
df_merge = df_germany_gdr[df_germany_gdr["Year"].isin(YEARS_TO_MERGE)].copy()
df_remaining = df_germany_gdr[~df_germany_gdr["Year"].isin(YEARS_TO_MERGE)].copy()

# Aggregate numeric data for merge years
df_aggregated = df_merge.groupby("Year", as_index=False).sum(numeric_only=True)

# Fix incorrect 1988 growth value (was summed as 0)
mask_1988 = (df_aggregated["Year"] == 1988) & (df_aggregated["Country Growth (%)"] == 0)
df_aggregated.loc[mask_1988, "Country Growth (%)"] = np.nan

# Assign unified partner name and keep required features
df_aggregated["Partner Name"] = "Germany"
df_aggregated = df_aggregated[FEATURES_TO_KEEP]

# Combine merged and remaining data
df_final = (
    pd.concat([df_aggregated, df_remaining])
    .sort_values(by="Year")
    .reset_index(drop=True)
)

# Preview final dataset
df_final.head()

Partner Name  Year  Country Growth (%)  AHS Simple Average (%)  \
0      Germany  1988                 NaN                   17.99   
1      Germany  1989                3.78                   23.74   
2      Germany  1990               12.71                   52.93   
3      Germany  1991                6.56                   14.67   
4      Germany  1992                2.46                   14.58   

   MFN Simple Average (%)  Export (US$ Thousand)  Import (US$ Thousand)  
0                   28.45           3.639411e+07           4.361709e+07  
1                   46.84           5.822599e+07           7.740300e+07  
2                   41.65           7.386752e+07           9.695782e+07  
3                   15.62           1.033760e+08           1.216899e+08  
4                   15.84           1.664047e+08           1.904864e+08

In [21]:
# Check data information
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Partner Name            34 non-null     object 
 1   Year                    34 non-null     int64  
 2   Country Growth (%)      33 non-null     float64
 3   AHS Simple Average (%)  34 non-null     float64
 4   MFN Simple Average (%)  34 non-null     float64
 5   Export (US$ Thousand)   34 non-null     float64
 6   Import (US$ Thousand)   34 non-null     float64
dtypes: float64(5), int64(1), object(1)
memory usage: 2.0+ KB


In [22]:
# Check for duplicates in the processed data
print(f"Number of duplicate records: {df_final.duplicated().sum()}")

Number of duplicate records: 0


In [23]:
# Count the number of null (missing) values in each column
df_final.isnull().sum()

Partner Name              0
Year                      0
Country Growth (%)        1
AHS Simple Average (%)    0
MFN Simple Average (%)    0
Export (US$ Thousand)     0
Import (US$ Thousand)     0
dtype: int64

In [24]:
# Save the processed data
df_final.to_csv("../data/processed/germany_trade_data.csv", index=False)

**Observations**

After merging the records for the **German Democratic Republic** and **Germany** for the years 1988, 1989, and 1990, we observed that the `Country Growth (%)` record for the year 1988 is missing in the merged data. This is because the `Country Growth (%)` was missing for both the German Democratic Republic and Germany in the year 1988, resulting in a single null value in the `Country Growth (%)` column of our new DataFrame. It was also observed that the new DataFrame contains no duplicate records, and the column datatypes are correct.

We will **not** be removing this null value because dropping a record in a time-series dataset can introduce a break in the data and a potential bias in the analysis. Leaving the missing value as `NaN` (Not a Number) is the most accurate way to represent the fact that the data is truly unavailable for that specific time point. Most data analysis and visualization tools are designed to handle these missing values successfully.

We have also saved our processed data to the appropriate directory. In the next section, we will load our [processed data](../data/processed/germany_trade_data.csv) and begin our exploratory data analysis and visualization.